### ISSUE TITLE
On dealing with shared holders

### ISSUE THREAD

**Issue by SylvainCorlay** [QuantStack/xwidgets#150]
The situation where we need shared holders occurs when a function needs to return a complex compound widget that reuses some widget in multiple cases (no clear ownership).

In most cases, users of `xwidgets` don't need to know about holders. It should probably not be the case either for shared holders.

This is a bit too verbose:

```cpp
auto box_of_buttons() {
    auto b1 = xw::button_generator()
        .description("button 1")
        .finalize();
    auto b2 = xw::button_generator()
        .description("button 2")
        .finalize();
    auto h1 = xw::make_shared_holder<xw::xwidget, xw::button>(std::move(b1));
    auto h2 = xw::make_shared_holder<xw::xwidget, xw::button>(std::move(b2));
    
    xw::hbox box;
    box.add(h1);   /// need a new add method with a holder to box.
    box.add(h1);   /// 
    box.add(h2);   /// 
    return box;
}
```

**Comment by JohanMabille** [comment]
I agree that we should find a way so the user does not have to deal with holders, however the sharing of a widget has to be explicitly specified.

A first solution could be to add an overload taking a shared pointer, so we can type the following:

```cpp
auto box_of_buttons() {
    auto b1 = xw::button_generator()
        .description("button 1")
        .finalize();
    auto b2 = xw::button_generator()
        .description("button 2")
        .finalize();
    auto shb1 = std::make_shared<xw::button>(std::move(b1));

    xw::hbox box;
    box.add(shb1);
    box.add(shb1);
    box.add(std::make_shared<xw::button>(std::move(b2)));
```

Another solution is to add a "add_and_share" method to the box, that will build the holder for the user and returns it so the user can add it to other containers; this requires a second overload taking a `xholder` parameter.

```cpp
auto box_of_buttons() {
    auto b1 = xw::button_generator()
        .description("button 1")
        .finalize();
    auto b2 = xw::button_generator()
        .description("button 2")
        .finalize();
    
    xw::hbox box;
    auto shb1 = box.add_and_share(std::move(b1));
    box.add(shb1);
```

Notice that in both cases we'll have to specify internally the type for the ` CRTP` parameter of the `xholder`.

**Comment by JohanMabille** [comment]
The first solution has been implemented in #151.
